File: memory/consolidator.go
package memory

import (
	"context"
	"fmt"
	"log"
	"sort"
	"strings"
	"sync"
	"time"

	"memg/embed"
	"memg/llm"
	"memg/store"
)

// Consolidator periodically clusters old, low-significance event facts into
// pattern facts, marks the originals as historical, and preserves a summary
// of repeated behavior.
type Consolidator struct {
	repo      store.Repository
	provider  llm.Provider
	embedder  embed.Embedder
	interval  time.Duration
	stopCh    chan struct{}
	done      chan struct{}
	startOnce sync.Once
}

// NewConsolidator creates a consolidator that runs every interval.
// A zero interval disables automatic consolidation.
func NewConsolidator(repo store.Repository, provider llm.Provider, embedder embed.Embedder, interval time.Duration) *Consolidator {
	return &Consolidator{
		repo:     repo,
		provider: provider,
		embedder: embedder,
		interval: interval,
		stopCh:   make(chan struct{}),
		done:     make(chan struct{}),
	}
}

// Start begins the background consolidation loop. It is safe to call Start
// multiple times; subsequent calls are no-ops.
func (c *Consolidator) Start() {
	c.startOnce.Do(func() {
		if c.interval <= 0 || c.provider == nil || c.embedder == nil {
			close(c.done)
			return
		}
		go c.loop()
	})
}

// Stop signals the consolidator to stop and waits for the loop to exit.
func (c *Consolidator) Stop() {
	select {
	case <-c.stopCh:
	default:
		close(c.stopCh)
	}
	<-c.done
}

func (c *Consolidator) loop() {
	defer close(c.done)
	ticker := time.NewTicker(c.interval)
	defer ticker.Stop()

	for {
		select {
		case <-c.stopCh:
			return
		case <-ticker.C:
			c.consolidate()
		}
	}
}

// ConsolidateEntity triggers consolidation for a single entity. This allows
// callers to run consolidation on demand rather than waiting for the periodic
// background sweep.
func (c *Consolidator) ConsolidateEntity(ctx context.Context, entityUUID string) error {
	if c.provider == nil || c.embedder == nil {
		return fmt.Errorf("consolidator: provider and embedder are required")
	}
	return c.consolidateEntity(ctx, entityUUID)
}

// consolidate iterates over all known entities and consolidates old event
// facts for each one individually.
func (c *Consolidator) consolidate() {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
	defer cancel()

	entityUUIDs, err := c.repo.ListEntityUUIDs(ctx, 1000)
	if err != nil {
		log.Printf("memg consolidator: list entities: %v", err)
		return
	}

	for _, entityUUID := range entityUUIDs {
		select {
		case <-c.stopCh:
			return
		case <-ctx.Done():
			return
		default:
		}
		if err := c.consolidateEntity(ctx, entityUUID); err != nil {
			log.Printf("memg consolidator: entity %s: %v", entityUUID, err)
		}
	}
}

// consolidateEntity finds clusters of old event facts for a single entity and
// merges them into pattern facts.
func (c *Consolidator) consolidateEntity(ctx context.Context, entityUUID string) error {
	// Find old event facts (older than 30 days) for this entity.
	cutoff := time.Now().Add(-30 * 24 * time.Hour)

	filter := store.FactFilter{
		Types:               []store.FactType{store.FactTypeEvent},
		Statuses:            []store.TemporalStatus{store.TemporalCurrent},
		ExcludeExpired:      true,
		ReferenceTimeBefore: &cutoff,
		MaxSignificance:     store.SignificanceMedium,
	}

	facts, err := c.repo.ListFactsFiltered(ctx, entityUUID, filter, 500)
	if err != nil {
		return fmt.Errorf("load candidates: %w", err)
	}

	if len(facts) < 3 {
		return nil // Not enough facts to consolidate.
	}

	byTag := make(map[string][]*store.Fact)
	for _, f := range facts {
		tag := f.Tag
		if tag == "" {
			tag = "_untagged"
		}
		byTag[tag] = append(byTag[tag], f)
	}

	for tag, group := range byTag {
		if len(group) < 3 {
			continue // Need at least 3 similar facts to form a pattern.
		}

		// Sort by content for deterministic processing.
		sort.Slice(group, func(i, j int) bool {
			return group[i].Content < group[j].Content
		})

		// Ask the LLM to summarize the cluster into a pattern fact.
		var contents strings.Builder
		for _, f := range group {
			contents.WriteString("- ")
			contents.WriteString(f.Content)
			contents.WriteByte('\n')
		}

		prompt := fmt.Sprintf(`Summarize these %d related events into a single behavioral pattern statement.
The pattern should describe a recurring behavior or tendency.
Return ONLY the pattern statement, nothing else.
If these events don't form a meaningful pattern, respond with exactly: NONE

Events:
%s`, len(group), contents.String())

		req := &llm.Request{
			Messages:  []*llm.Message{llm.UserMessage(prompt)},
			MaxTokens: 100,
		}

		resp, err := c.provider.Chat(ctx, req)
		if err != nil {
			log.Printf("memg consolidator: llm call for entity %s tag %q: %v", entityUUID, tag, err)
			continue
		}

		pattern := strings.TrimSpace(resp.Content)
		if pattern == "" || strings.EqualFold(pattern, "NONE") {
			continue
		}

		vectors, err := c.embedder.Embed(ctx, []string{pattern})
		if err != nil || len(vectors) == 0 {
			log.Printf("memg consolidator: embed pattern for entity %s tag %q: %v", entityUUID, tag, err)
			continue
		}

		patternFact := &store.Fact{
			Content:        pattern,
			Embedding:      vectors[0],
			Type:           store.FactTypePattern,
			TemporalStatus: store.TemporalCurrent,
			Significance:   store.SignificanceMedium,
			Tag:            tag,
			ContentKey:     store.DefaultContentKey(pattern),
		}

		if err := c.repo.InsertFact(ctx, entityUUID, patternFact); err != nil {
			log.Printf("memg consolidator: insert pattern for entity %s tag %q: %v", entityUUID, tag, err)
			continue
		}

		// Mark originals as historical only after the pattern is persisted.
		for _, f := range group {
			if err := c.repo.UpdateTemporalStatus(ctx, f.UUID, store.TemporalHistorical); err != nil {
				log.Printf("memg consolidator: mark historical %s: %v", f.UUID, err)
			}
		}

		log.Printf("memg consolidator: created pattern from %d events (entity=%s tag=%s): %s", len(group), entityUUID, tag, pattern)
	}

	return nil
}

File: src/consolidator.ts
/**
 * Consolidator: clusters old, low-significance event facts into pattern facts.
 * Matches the Go memory/consolidator.go logic.
 */

import { randomUUID } from 'crypto';
import type { Embedder } from './embedder.js';
import { cosineSimilarity } from './search.js';
import { defaultContentKey, type Store } from './store.js';
import type { Fact, FactFilter } from './types.js';

/** Configurable thresholds for entity consolidation. */
export interface ConsolidateOptions {
  /** Minimum age in days for facts to be eligible. Default: 30. */
  minAgeDays?: number;
  /** Maximum significance for eligible facts. Default: 5. */
  maxSignificance?: number;
  /** Minimum facts per tag group to trigger consolidation. Default: 3. */
  minGroupSize?: number;
  /** Maximum facts to scan. Default: 500. */
  scanLimit?: number;
}

export interface ConsolidatorConfig {
  store: Store;
  embedder: Embedder;
  llmChat: (prompt: string) => Promise<string>;
  intervalMs?: number;
}

/**
 * Consolidate old event facts for a single entity into pattern facts.
 *
 * Logic (matches Go consolidateEntity):
 * 1. Query event facts older than `minAgeDays`, current status, significance <= `maxSignificance`
 * 2. Group by tag
 * 3. For groups with `minGroupSize`+ facts, ask LLM to produce a behavioral pattern
 * 4. If a semantically similar pattern already exists (>0.88), reinforce it instead of duplicating
 * 5. Otherwise embed the pattern, insert as pattern fact
 * 6. Mark originals as historical
 * 7. Return count of patterns created
 */
export async function consolidateEntity(
  store: Store,
  embedder: Embedder,
  llmChat: (prompt: string) => Promise<string>,
  entityUuid: string,
  opts?: ConsolidateOptions
): Promise<number> {
  const minAgeDays = opts?.minAgeDays ?? 30;
  const maxSignificance = opts?.maxSignificance ?? 5;
  const minGroupSize = opts?.minGroupSize ?? 3;
  const scanLimit = opts?.scanLimit ?? 500;

  const cutoff = new Date(Date.now() - minAgeDays * 24 * 60 * 60 * 1000).toISOString();

  const filter: FactFilter = {
    types: ['event'],
    statuses: ['current'],
    excludeExpired: true,
    referenceTimeBefore: cutoff,
    maxSignificance: maxSignificance,
  };

  const facts = await store.listFactsFiltered(entityUuid, filter, scanLimit);
  if (facts.length < minGroupSize) return 0;

  // Pre-load existing pattern facts for duplicate detection.
  const existingPatterns = await store.listFactsFiltered(
    entityUuid,
    { types: ['pattern'], statuses: ['current'] },
    scanLimit
  );

  const byTag = new Map<string, Fact[]>();
  for (const f of facts) {
    const tag = f.tag || '_untagged';
    const group = byTag.get(tag) ?? [];
    group.push(f);
    byTag.set(tag, group);
  }

  let patternsCreated = 0;

  for (const [tag, group] of byTag) {
    if (group.length < minGroupSize) continue;

    group.sort((a, b) => a.content.localeCompare(b.content));

    const contents = group.map((f) => `- ${f.content}`).join('\n');

    const prompt = `Summarize these ${group.length} related events into a single behavioral pattern statement.
The pattern should describe a recurring behavior or tendency.
Return ONLY the pattern statement, nothing else.
If these events don't form a meaningful pattern, respond with exactly: NONE

Events:
${contents}`;

    let pattern: string;
    try {
      pattern = await llmChat(prompt);
    } catch (err) {
      console.warn('[memg] consolidator: LLM call failed:', err);
      continue;
    }

    pattern = pattern.trim();
    if (!pattern || pattern.toUpperCase() === 'NONE') continue;

    let vectors: number[][];
    try {
      vectors = await embedder.embed([pattern]);
    } catch (err) {
      console.warn('[memg] consolidator: embed failed:', err);
      continue;
    }
    if (vectors.length === 0) continue;

    const newEmbedding = vectors[0];

    // Check for duplicate pattern: if an existing pattern is semantically
    // very similar (>0.88), reinforce it instead of creating a duplicate.
    let duplicateFound = false;
    for (const existing of existingPatterns) {
      if (!existing.embedding || existing.embedding.length === 0) continue;
      if (existing.embedding.length !== newEmbedding.length) continue;
      const sim = cosineSimilarity(newEmbedding, existing.embedding);
      if (sim > 0.88) {
        try {
          await store.reinforceFact(existing.uuid, null);
        } catch (err) {
          console.warn('[memg] consolidator: reinforce existing pattern failed:', err);
        }
        duplicateFound = true;
        break;
      }
    }

    if (duplicateFound) {
      // Still mark originals as historical even when reinforcing an existing pattern.
      for (const f of group) {
        try {
          await store.updateTemporalStatus(f.uuid, 'historical', 'consolidated');
        } catch (err) {
          console.warn('[memg] consolidator: updateTemporalStatus failed:', err);
        }
      }
      continue;
    }

    const patternFact: Fact = {
      uuid: randomUUID(),
      content: pattern,
      embedding: newEmbedding,
      factType: 'pattern',
      temporalStatus: 'current',
      significance: maxSignificance,
      contentKey: defaultContentKey(pattern),
      tag,
      slot: '',
      confidence: 1.0,
      embeddingModel: embedder.modelName(),
      sourceRole: '',
      reinforcedCount: 0,
      recallCount: 0,
    };

    try {
      await store.insertFact(entityUuid, patternFact);
    } catch (err) {
      console.warn('[memg] consolidator: insertFact failed:', err);
      continue;
    }

    // Add the newly inserted pattern to the existing patterns list so
    // subsequent tag groups can detect duplicates against it.
    existingPatterns.push(patternFact);

    for (const f of group) {
      try {
        await store.updateTemporalStatus(f.uuid, 'historical', patternFact.uuid);
      } catch (err) {
        console.warn('[memg] consolidator: updateTemporalStatus failed:', err);
      }
    }

    patternsCreated++;
  }

  return patternsCreated;
}
